src/lib/tls/tls_session.cpp

```diff
@@ -90,9 +145,27 @@ Session::Session(const uint8_t ber[], size_t ber_len)
         .decode(server_service)
         .decode(server_port)
         .decode(srp_identifier_str)
-        .decode(srtp_profile)
-      .end_cons()
-      .verify_end();
+        .decode(srtp_profile);
+
+   m_version = Protocol_Version(major_version, minor_version);
+
+   if(struct_version == TLS_SESSION_PARAM_STRUCT_VERSION_TLS12 &&
+      !m_version.is_pre_tls_13())
+      {
+      throw Decoding_Error("Serialized TLS session scheme does not match the protocol version expectations");
+      }
+
+   if(struct_version == TLS_SESSION_PARAM_STRUCT_VERSION_TLS13)
+      {
+      seqdec
+         .decode(m_early_data_allowed)
+         .decode_integer_type(m_max_early_data_bytes)
+         .decode_integer_type(m_ticket_age_add)
+         .decode_integer_type(m_lifetime_hint);
+      }
+
+   seqdec.end_cons();
+   dec.verify_end();
```

Comment:
This is compatibility code to allow the new session implementation to read legacy session objects and resume sessions that were saved before migrating to Botan 3.0.

It introduces quite a bit of complexity due to handling the `BER_Decoder` objects properly. Is this even necessary? I feel it might be okay that Botan 2.x sessions are lost when updating to Botan 3.0?

Reply:
dropped.